### PR TITLE
fix(azureopenai): align provider with GA v1 contracts

### PR DESCRIPTION
**Pull request by IceCodeNew**
## Description

Use the public OpenAI client with Azure's GA `/openai/v1/` endpoint, preserving explicit endpoint and credential precedence.

- Accept an API key, static Microsoft Entra token, or sync/async token provider.
- Keep header generation, token refresh, retries and redirect handling in the SDK.
- Accept explicit `api_version="v1"` and `OPENAI_API_VERSION=v1` as the GA route default. Reject dated versions and deployment-in-path options. Preserve valid `default_query["api-version"]="v1"`.
- Advertise the implemented GA capabilities; keep preview media and Batch disabled.

The explicit `v1` acceptance fixes a constructor rejection found during independent review. It does not add a legacy route or change the service version. Microsoft defines the optional API-version query default as `v1` in the [official schema](https://github.com/Azure/azure-rest-api-specs/blob/3e3f840f52f06cc593b6301d30ea0e04802473b2/specification/ai/data-plane/OpenAI.v1/azure-v1-v1-generated.json).

The provider owns Azure configuration; shared OpenAI code owns request, response and stream conversion. The two constructor overrides prevent ambient credentials or endpoints from overriding explicit Azure options. No SDK dependency change is needed for this correction.

## Verification

The new explicit-version regression failed before the fix. With the fix:

- Azure provider tests: 19 passed.
- Full unit suite in this worktree's declared dependency environment, Python 3.13: 2442 passed, 69 skipped.
- Changed-file Ruff lint and format checks passed. The all-stable scan reports only test assertions, the intentional test namespace package, and COM812's conflict with the formatter.
- After `uv sync --all-extras --python 3.13`, `UV_NO_SYNC=1 uv run pre-commit run --all-files` passed, including mypy over 287 source files. Preserving the synced environment prevents a nested `uv run` from pruning optional SDK dependencies.
- `git diff --check` passed.

Local wire tests exercise the real SDK with a mock transport. They are not live Azure verification. Live authentication across Azure resource families, exact-head CodeRabbit review and final provider-level acceptance remain pending; this PR stays Draft.

## AI usage

Generated with Amp using GPT-5.6 Sol X-HIGH.

### CODE REVIEW

File: src/any_llm/providers/azureopenai/azureopenai.py [hunk 1]
```diff
@@ -46,15 +73,53 @@ def _init_client(
...
+        default_headers = dict(response_kwargs.pop("default_headers", None) or {})
+        if self._responses_api_key_auth:
+            default_headers["api-key"] = api_key
+        if default_headers:
+            response_kwargs["default_headers"] = default_headers
+
+        self.responses_client = AsyncOpenAI(
+            api_key=response_api_key,
+            base_url=f"{azure_endpoint.rstrip('/')}/openai/v1/",
+            **response_kwargs,
```
[the remark this answers is not in the document]
Comment (IceCodeNew) [reply src/any_llm/providers/azureopenai/azureopenai.py]: Correction after the current official-document audit: the earlier conclusion should not be relied on. Azure OpenAI credentials now require an HTTPS endpoint. The final provider tests reject HTTP and missing-scheme endpoints, and remove api-key and Authorization credentials from cross-origin or downgraded redirects.

File: src/any_llm/providers/azureopenai/azureopenai.py [hunk 2]
```diff
@@ -44,17 +73,69 @@ def _init_client(
...
+            default_query=default_query,
             **kwargs,
         )
+        # Azure's current image and audio reference keeps these operations on v1 preview:
+        # https://learn.microsoft.com/azure/foundry/openai/reference-preview-latest
+        self._media_client = AsyncOpenAI(
+            api_key=client_api_key,
+            base_url=base_url,
+            default_query={**default_query, "api-version": "preview"},
+            **kwargs,
```
[the remark this answers is not in the document]
Comment (IceCodeNew) [reply src/any_llm/providers/azureopenai/azureopenai.py]: Correction after the current official-document audit: the earlier GA bearer-key conclusion was wrong. Azure API keys now use the documented api-key header on GA and preview routes; only Microsoft Entra credentials use Authorization bearer. AsyncOpenAI remains necessary to preserve the current /openai/v1 media paths, with request hooks enforcing header selection and redirect stripping.

**Comment by IceCodeNew** [comment]
@coderabbitai review

File: pyproject.toml [hunk 3]
```diff
@@ -10,7 +10,7 @@ requires-python = ">=3.11"
 version = "0.0.0-dev" # There is a regex in the release.yaml file that looks for this version string to set the version in the pyproject.toml file. Do not change this version string without updating the regex.
 dependencies = [
   "pydantic>2,<3",
-  "openai>=2.18.0",
+  "openai>=2.47.0",
```
[the remark this answers is not in the document]
Comment (IceCodeNew) [reply pyproject.toml]: The implementation authority is current `openai` 3.6.0, while the dependency floor is the first formal release exposing every required public API. Release-by-release construction tests show 2.46.0 rejects the provider HTTPX2 client and 2.47.0 accepts it with provider tests and strict mypy passing. Raising the floor further would exclude working releases without adding a required contract.

File: tests/unit/providers/test_azureopenai_provider.py [hunk 4]
```diff
@@ -1,126 +1,294 @@
...
+    async with httpx2.AsyncClient(transport=httpx2.MockTransport(handler)) as http_client:
+        provider = AzureopenaiProvider(
+            api_key="azure-key",
+            api_base=endpoint,
+            api_version="2024-06-01",
+            azure_deployment="ignored-for-v1",
+            default_query={"api-version": "legacy", "trace": "enabled"},
+            http_client=http_client,
+        )
+        await provider._aresponses(ResponsesParams(model="deployment", input="hello"))
```
[the remark this answers is not in the document]
Comment (IceCodeNew) [reply tests/unit/providers/test_azureopenai_provider.py]: Resolved on the current head. Azure transport tests cover Pydantic responses.parse() and dataclass responses.create() plus parse_responses_output. Both exercise the HTTPX2 mock transport and assert the current v1 route, api-key authentication for Azure keys, omission of api-version, JSON Schema request bodies, and parsed result types.

**Comment by IceCodeNew** [comment]
@coderabbitai review

File: tests/unit/providers/test_azureopenai_provider.py [hunk 5]
```diff
@@ -1,126 +1,375 @@
...
+        return httpx2.Response(200, json={"data": [], "object": "list"})
+
+    async with httpx2.AsyncClient(transport=httpx2.MockTransport(handler)) as http_client:
+        provider = AzureopenaiProvider(
+            api_key="azure-key",
+            api_base="https://resource.openai.azure.com",
+            default_query={"api-version": "legacy", "trace": "enabled"},
+            http_client=http_client,
+        )
+        await provider.client.models.list()
```
[the remark this answers is not in the document]
Comment (IceCodeNew) [reply tests/unit/providers/test_azureopenai_provider.py]: Resolved on the current head. The v1 routing test calls `provider.alist_models()` with a representative Azure model payload and asserts the normalized any-llm `Model`, covering inherited model-list conversion together with the GA route and authentication.

**Comment by IceCodeNew** [comment]
@coderabbitai review

**Comment by IceCodeNew** [comment]
@coderabbitai resume

File: tests/unit/providers/test_azureopenai_provider.py [hunk 6]
```diff
@@ -1,126 +1,414 @@
...
-        mock_client_instance.chat.completions.create = AsyncMock(return_value=mock_response)
 
-        yield mock_client_instance, mock_azure_client
+@pytest.mark.asyncio
+@pytest.mark.parametrize(
+    ("client_kwargs", "expected_api_version"),
+    [
+        ({"api_version": "preview"}, "preview"),
+        ({"default_query": {"api-version": "v1"}}, "v1"),
+    ],
```
[the remark this answers is not in the document]
Comment (IceCodeNew) [reply tests/unit/providers/test_azureopenai_provider.py:144]: Resolved on the current head. The wire test covers both inputs together and verifies that explicit `api_version="preview"` takes precedence over `default_query["api-version"]`.

File: src/any_llm/providers/azureopenai/azureopenai.py [hunk 7]
```diff
@@ -44,17 +80,64 @@ def _init_client(
...
+        else:
+            default_query.pop("api-version", None)
+        azure_ad_token = kwargs.pop("azure_ad_token", None)
+        azure_ad_token_provider = kwargs.pop("azure_ad_token_provider", None)
+
+        # Explicit credentials take precedence over environment credentials.
+        if (azure_ad_token is not None or azure_ad_token_provider is not None) and self._api_key_from_environment:
+            api_key = None
+        elif azure_ad_token is None and azure_ad_token_provider is None and self._api_key_from_environment:
+            azure_ad_token = os.getenv("AZURE_OPENAI_AD_TOKEN")
```
[the remark this answers is not in the document]
Comment (IceCodeNew) [reply src/any_llm/providers/azureopenai/azureopenai.py]: Fixed in signed commit b13f7ace3ebea1b1000c6095f333fb69e8b9ebf7. An empty AZURE_OPENAI_API_KEY environment value now normalizes to None, while an explicit empty api_key argument retains its existing no-fallback behavior. The regression test proves an environment Entra token authenticates alone. The Azure suite passes 25 tests, and the all-files pre-commit run passes.

**Comment by IceCodeNew** [comment]
@coderabbitai resume

File: src/any_llm/providers/azureopenai/azureopenai.py [hunk 8]
```diff
@@ -1,41 +1,106 @@
...
+        message = f"Azure OpenAI endpoint must include a hostname, got {endpoint!r}."
+        raise ValueError(message)
+    origin_hostname = origin.hostname
+    origin_port = origin.port or 443
+
+    async def apply_api_key(request: Request) -> None:
+        request_port = request.url.port or (443 if request.url.scheme == "https" else 80)
+        request.headers.pop("Authorization", None)
+        if (request.url.scheme, request.url.host, request_port) == ("https", origin_hostname, origin_port):
+            request.headers["api-key"] = api_key
```
[the remark this answers is not in the document]
Comment (IceCodeNew) [reply src/any_llm/providers/azureopenai/azureopenai.py]: This finding is factually incorrect for Azure OpenAI API-key authentication, so the implementation remains unchanged. The current Microsoft GA OpenAPI schema at Azure/azure-rest-api-specs commit a6943a926f76b3a2f90371b3466157eddf760e25 defines AzureApiKey as an apiKey scheme in header api-key and AzureOAuth2 as the bearer alternative; GA Responses and Chat operations accept either scheme. The provider therefore sends api-key for static Azure keys and Authorization bearer for Entra tokens. The lifecycle page has an OpenAI(api_key=...) SDK example, but it does not override the published Azure wire schema. The source and conflict are now pinned in the PR body; redirect tests cover both credential forms.

File: tests/unit/providers/test_azureopenai_provider.py [hunk 9]
```diff
@@ -1,126 +1,618 @@
...
+            http_client=http_client,
+        )
+        models = await provider.client.models.list()
+
+    assert models.data == []
+    assert requests[0].headers["api-key"] == "azure-key"
+    assert "authorization" not in requests[0].headers
+    assert "api-key" not in requests[1].headers
+    assert "authorization" not in requests[1].headers
+    assert len(requests) == 2
```
[the remark this answers is not in the document]
Comment (IceCodeNew) [reply tests/unit/providers/test_azureopenai_provider.py]: Covered in signed commit 6a1f65df. Static azure_ad_token and asynchronous azure_ad_token_provider cases both assert bearer Authorization on the first request and no Authorization on the cross-origin redirect. HTTPX2 already performs the standards-required stripping, so no redundant provider hook was added. The focused Azure and shared Batch command passes 43 tests, and the exact-head changed-file command passes 364 tests with 41 optional-dependency skips.

File: tests/unit/providers/test_openai_batch_results.py [hunk 10]
```diff
@@ -31,6 +31,7 @@ async def test_aretrieve_batch_results_success() -> None:
         mock_batch = MagicMock()
         mock_batch.status = "completed"
         mock_batch.output_file_id = "file-output-123"
+        mock_batch.error_file_id = None
```
[the remark this answers is not in the document]
Comment (IceCodeNew) [reply tests/unit/providers/test_openai_batch_results.py]: Covered in signed commit 20c53888. The shared Batch success test now gives the completed batch distinct output_file_id and error_file_id values, verifies both file IDs are fetched, preserves the successful item, and parses the separate error-file entry as BatchResultItem.error. The exact-head changed-file command passes 364 tests with 41 optional-dependency skips.

**Comment by IceCodeNew** [comment]
@coderabbitai full review

File: docs/quickstart.md [hunk 11]
```diff
@@ -96,6 +96,12 @@ print(response.choices[0].message.content)
 
 **Finding model names:** Check the [providers page](providers.md) for provider IDs, or use the [`list_models`](api/list-models.md) API to see available models for your provider.
 
+### Azure OpenAI API modes
+
+The `azureopenai` provider uses Azure OpenAI's GA v1 base URL, `{AZURE_OPENAI_ENDPOINT}/openai/v1/`, for Responses, Chat Completions, embeddings, and model operations. Image generation, speech, and transcription use Azure's v1 preview routes with `api-version=preview`. The `model` value is your Azure deployment name. The provider preserves legacy `api_version="v1"` and `api_version="preview"` query values, ignores other version values, and accepts but ignores `azure_deployment`. Responses and Chat Completions are available when the selected deployment supports them.
```
[the remark this answers is not in the document]
Comment (IceCodeNew) [reply docs/quickstart.md]: Fixed in `66b53e42`. The Azure OpenAI quickstart now states that `api_base`, `azure_endpoint`, and `AZURE_OPENAI_ENDPOINT` must resolve to HTTPS, matching constructor validation. The changed-file pre-commit gate passes from a clean login shell using the Project pre-setup environment.

**Comment by IceCodeNew** [comment]
@coderabbitai full review

File: tests/unit/providers/test_openai_base_provider.py [hunk 12]
```diff
@@ -28,6 +32,110 @@ def test_timeout_capability_is_native_for_openai_compatible_providers() -> None:
...
+    await cleanup_started.wait()
+    operation.cancel()
+    finish_cleanup.set()
+
+    with pytest.raises(BaseExceptionGroup) as raised:
+        await operation
+
+    assert any(isinstance(error, RuntimeError) for error in raised.value.exceptions)
+    assert any(isinstance(error, asyncio.CancelledError) for error in raised.value.exceptions)
+    provider.client.files.delete.assert_awaited_once_with("file-input")
```
[the remark this answers is not in the document]
Comment (IceCodeNew) [reply tests/unit/providers/test_openai_base_provider.py]: Fixed in `15f0b1ae`. The cancellation test now sets a completion flag only after the blocked delete coroutine resumes, then asserts that flag after the operation returns. This proves cleanup completed rather than only proving that the mock was awaited. The targeted test and changed-file pre-commit pass.

**Comment by IceCodeNew** [comment]
@coderabbitai full review

Please review exact head `15f0b1aeaf8c828035257bf760669872eab24e16`. The latest test now proves file cleanup has completed before cancellation handling returns.

File: src/any_llm/providers/azureopenai/azureopenai.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (IceCodeNew) [reply src/any_llm/providers/azureopenai/azureopenai.py]: The reciprocal empty-environment case is fixed at exact signed head `3921769c66012a3c91750244d461b8e2730c2d4a`: `AZURE_OPENAI_AD_TOKEN=""` now normalizes to `None`, so a non-empty `AZURE_OPENAI_API_KEY` remains the sole credential. The direct regression passes, removing the normalization makes it fail with the mutual-exclusion error, and the existing non-empty credential conflict test still passes. The full Azure provider file passes 41 tests and focused pre-commit passes. This thread was already resolved.

**Comment by IceCodeNew** [comment]
@coderabbitai full review

Please review exact head `3921769c66012a3c91750244d461b8e2730c2d4a`, including the empty `AZURE_OPENAI_AD_TOKEN` normalization and its direct regression.

File: src/any_llm/providers/azureopenai/azureopenai.py [hunk 13]
```diff
@@ -1,41 +1,129 @@
...
+    """Apply one Azure provider's credentials after SDK caller headers merge."""
+
+    def __init__(self, *, azure_api_key: str | None, azure_origin: _AzureAuthOrigin, **kwargs: Any) -> None:
+        self._azure_api_key = azure_api_key
+        self._azure_origin = azure_origin
+        super().__init__(**kwargs)
+
+    @override
+    async def _prepare_request(self, request: Request) -> None:
+        await super()._prepare_request(request)
```
[the remark this answers is not in the document]
Comment (IceCodeNew) [reply src/any_llm/providers/azureopenai/azureopenai.py]: Valid concern on the old implementation. The rebuilt exact head `d7765f1c103451f625c8a96513ef834acb7fdb6a` deletes the private `_prepare_request` override instead of pinning its ordering. It now passes a public async credential callable to `AsyncOpenAI`, which owns refresh and request preparation. The real-client wire suite passes unchanged on the repository floor `openai==2.18.0` and current `openai==3.8.0`, including token refresh before an SDK retry. This removes the private-hook compatibility surface that the comment identified.

**Comment by IceCodeNew** [comment]
Correction to my earlier review replies and the CodeRabbit learning derived from them: the categorical claim that an Azure API key must use the `api-key` header for GA v1 was not supported by the current sources. [Microsoft’s current Python v1 example](https://learn.microsoft.com/azure/foundry/openai/api-version-lifecycle) passes the key to the generic OpenAI client as `api_key`, which sends bearer Authorization. The [current GA OpenAPI](https://github.com/Azure/azure-rest-api-specs/blob/a6943a926f76b3a2f90371b3466157eddf760e25/specification/ai/data-plane/OpenAI.v1/azure-v1-v1-generated.json) declares API key, Authorization, and OAuth schemes, while a Batch curl example still uses `api-key`.

The rebuilt head therefore removes the custom header and redirect implementation and delegates authentication to the public `AsyncOpenAI` contract. It passes real-client wire tests on `openai==2.18.0` and current `openai==3.8.0`. No Azure live credential was available, so acceptance across every Azure resource and endpoint family remains unverified rather than being stated as settled.

**Comment by IceCodeNew** [comment]
@coderabbitai full review

Please review exact head `d7765f1c103451f625c8a96513ef834acb7fdb6a`. This head rebuilds the PR from current Microsoft GA v1 documentation and the public OpenAI SDK contract, removes the old private hook and custom transport, and includes real-client wire tests on both `openai==2.18.0` and current `openai==3.8.0`. The PR body lists the remaining live-Azure gaps.

File: tests/unit/providers/test_azureopenai_provider.py [hunk 14]
```diff
@@ -1,154 +1,405 @@
...
-    with mock_azureopenai_provider() as (_, mock_azure_client):
-        AzureopenaiProvider(api_key=api_key, api_base=api_base)
+@pytest.mark.parametrize(
+    ("legacy_options", "expected_parameter"),
+    [
+        ({"api_version": "2024-10-21"}, "api_version"),
+        ({"azure_deployment": "deployment-name"}, "azure_deployment"),
+        ({"default_query": {"api-version": "2024-10-21"}}, "default_query['api-version']"),
+    ],
+)
```
[the remark this answers is not in the document]
Comment (IceCodeNew) [reply tests/unit/providers/test_azureopenai_provider.py:145]: Valid. Fixed in signed commit `4767ca33`: the existing constructor contract now passes `default_query={"trace": "1"}` and asserts that `AsyncOpenAI` receives the mapping unchanged. The dated `default_query["api-version"]` rejection cases remain intact. The focused Azure provider suite is 16 passed and configured Ruff passes.

**Comment by IceCodeNew** [comment]
@coderabbitai full review

Please review exact head `4767ca33bf00527cff8883f31b39666200fc25e5`. The only change after the prior full review folds the requested non-version `default_query` pass-through assertion into the existing constructor contract.

File: src/any_llm/providers/azureopenai/azureopenai.py [hunk 15]
```diff
@@ -1,62 +1,167 @@
...
+            "The `api_key`, `azure_ad_token` and `azure_ad_token_provider` arguments are mutually exclusive; "
+            "only one can be passed at a time."
+        )
+        raise OpenAIError(message)
+
+    if azure_ad_token_provider is not None:
+
+        async def get_token() -> str:
+            token = azure_ad_token_provider()
+            resolved_token = token if isinstance(token, str) else await token
```
[the remark this answers is not in the document]
Comment (IceCodeNew) [reply src/any_llm/providers/azureopenai/azureopenai.py]: Addressed in [`b26c77c1`](https://github.com/IceCodeNew/any-llm/commit/b26c77c1df83718b4a38acce3abb49b5a5ffb7fc). The synchronous provider now runs through `asyncio.to_thread`; if it returns an awaitable, that awaitable is consumed on the caller event loop. `test_azureopenai_refreshes_sync_entra_provider_before_sdk_retry` verifies both refresh attempts execute outside the event-loop thread and produce distinct Authorization values. The focused contract passes on OpenAI Python 2.18.0, 3.3.1, and 3.8.0.

File: src/any_llm/providers/azureopenai/azureopenai.py [hunk 16]
```diff
@@ -1,62 +1,167 @@
...
-            azure_endpoint=azure_endpoint,
-            api_version=api_version,
-            azure_ad_token=azure_ad_token,
+        # Microsoft v1 uses an implicit API version and the deployment name in
+        # `model`. Failing legacy options explicitly prevents a dated-route
+        # configuration from appearing to work while being silently discarded.
+        # https://learn.microsoft.com/azure/foundry/openai/api-version-lifecycle
+        if api_version is not None or os.getenv("OPENAI_API_VERSION"):
+            parameter_name = "api_version"
+            raise UnsupportedParameterError(parameter_name, self.PROVIDER_NAME)
```
[the remark this answers is not in the document]
Comment (IceCodeNew) [reply src/any_llm/providers/azureopenai/azureopenai.py]: Addressed in [`b26c77c1`](https://github.com/IceCodeNew/any-llm/commit/b26c77c1df83718b4a38acce3abb49b5a5ffb7fc). Explicit `api_version` and `OPENAI_API_VERSION` now have separate rejection branches, so the exception names the source that the caller must remove. `test_azureopenai_rejects_legacy_api_version_environment` asserts the environment-variable name.

File: src/any_llm/providers/azureopenai/azureopenai.py [hunk 17]
```diff
@@ -1,62 +1,167 @@
...
+            message = f"Azure OpenAI batch completion_window must be '24h', got {completion_window!r}."
+            raise InvalidRequestError(message, provider_name=self.PROVIDER_NAME)
+
+        # Azure's current guide requires this create-time endpoint even when the
+        # JSONL file contains `/v1/responses`; the service inspects each line to
+        # select Chat Completions or Responses.
+        # https://learn.microsoft.com/azure/foundry/openai/how-to/batch
+        return await super()._acreate_batch(
+            input_file_path,
+            "/chat/completions",
```
[the remark this answers is not in the document]
Comment (IceCodeNew) [reply src/any_llm/providers/azureopenai/azureopenai.py]: Not changed after checking the current Microsoft sources. The newer August 2026 [official Python Batch guide at immutable commit `10b7fc9e`](https://github.com/MicrosoftDocs/azure-ai-docs/blob/10b7fc9e6186aabe5ee440bdccaa4c2f41d68faf/articles/foundry/openai/includes/batch/batch-python.md#L206-L227) calls `batches.create(endpoint="/chat/completions")` and states that the service reads the JSONL to distinguish Chat Completions from Responses. The older June GA schema lists `/v1/chat/completions`, so the sources conflict. Following the newer official guide, the provider keeps `/chat/completions` as the create-time placeholder while accepting `/v1/chat/completions` and `/v1/responses` inside JSONL. `test_azureopenai_maps_batch_input_routes_to_documented_create_route` fixes both sides of that contract. Live Azure verification remains unavailable and is stated as an open gap.

File: tests/unit/providers/test_azureopenai_provider.py [hunk 18]
```diff
@@ -1,154 +1,406 @@
...
+        AzureopenaiProvider(api_key="key")
+    with pytest.raises(MissingApiKeyError, match="AZURE_OPENAI_API_KEY or AZURE_OPENAI_AD_TOKEN"):
+        AzureopenaiProvider(api_base="https://resource.openai.azure.com")
+    token = "token"  # noqa: S105
+    with pytest.raises(OpenAIError, match="mutually exclusive"):
+        AzureopenaiProvider(
+            api_key="key",
+            api_base="https://resource.openai.azure.com",
+            azure_ad_token=token,
+        )
```
[the remark this answers is not in the document]
Comment (IceCodeNew) [reply tests/unit/providers/test_azureopenai_provider.py:197]: Addressed in [`143334f8`](https://github.com/IceCodeNew/any-llm/commit/143334f8b57aeae543aa3882b41c4c8a0536f4ed). The parameterized environment test covers both empty-value directions, the constructor test covers API key plus token-provider exclusivity, and the async wire test proves an empty provider token fails before transport. These run through the real `AsyncOpenAI` client where a request is required.

File: tests/unit/providers/test_azureopenai_provider.py [hunk 19]
```diff
@@ -1,154 +1,406 @@
...
+            response = await provider.aresponses(model="deployment-name", input_data="Hello")
+        finally:
+            await provider.client.close()
 
-    with pytest.raises(MissingApiKeyError, match="AZURE_OPENAI_API_KEY"):
-        AzureopenaiProvider(api_key=None, api_base="https://test.openai.azure.com")
+    assert response.id == "resp-test"
+    assert paths == ["/openai/v1/responses"]
+    assert headers == ["Bearer api-key"]
+    assert json.loads(bodies[0]) == {"input": "Hello", "model": "deployment-name"}
```
[the remark this answers is not in the document]
Comment (IceCodeNew) [reply tests/unit/providers/test_azureopenai_provider.py]: Addressed in [`143334f8`](https://github.com/IceCodeNew/any-llm/commit/143334f8b57aeae543aa3882b41c4c8a0536f4ed). One parameterized real-client contract covers the Pydantic `responses.parse()` path and the separate dataclass `parse_responses_output` path. Both assert `/openai/v1/responses`, JSON Schema serialization, and normalized parsed output.

**Comment by IceCodeNew** [comment]
@coderabbitai full review

Please review exact head `143334f8b57aeae543aa3882b41c4c8a0536f4ed`. This rebuild follows Microsoft GA v1 documentation and the public `AsyncOpenAI` contract, preserves optional `api-version=v1`, separates Batch JSONL endpoints from the newer guide’s create-time `/chat/completions` placeholder, and covers sync/async credentials, errors, retries, structured Responses, timeout, and cancellation with a real SDK client. The PR body lists unresolved live-Azure evidence.

File: tests/unit/providers/test_azureopenai_provider.py [hunk 20]
```diff
@@ -1,154 +1,510 @@
...
+            elif b'"stream":true' in body:
+                self.send_response(_HTTP_OK)
+                self.send_header("Content-Type", "text/event-stream")
+                self.send_header("Content-Length", str(len(_CHAT_STREAM)))
+                self.end_headers()
+                self._write_body(_CHAT_STREAM)
+            else:
+                self._send_json(_HTTP_OK, _CHAT_RESPONSE)
+
+        def do_POST(self) -> None:
```
[the remark this answers is not in the document]
Comment (IceCodeNew) [reply tests/unit/providers/test_azureopenai_provider.py]: The finding no longer applies on exact head `0e95485c77fee5c4258f0bec60e79f125e2dc88b`. The final ablation replaced the custom `BaseHTTPRequestHandler` server with `httpx.MockTransport`, so `_Handler.do_POST` and its dynamic override surface no longer exist. The same 15 wire contracts pass on OpenAI Python 2.18.0 and 3.8.0, and all-files strict mypy passes.

**Comment by IceCodeNew** [comment]
@coderabbitai full review

File: src/any_llm/providers/azureopenai/azureopenai.py [hunk 21]
```diff
@@ -1,62 +1,147 @@
...
+            "only one can be passed at a time."
+        )
+        raise OpenAIError(message)
+
+    if azure_ad_token_provider is not None:
+
+        async def get_token() -> str:
+            token = await asyncio.to_thread(azure_ad_token_provider)
+            resolved_token = token if isinstance(token, str) else await token
+            if not resolved_token:
```
[the remark this answers is not in the document]
Comment (IceCodeNew) [reply src/any_llm/providers/azureopenai/azureopenai.py]: Fixed in signed commit `26fd1522029c92c0476990172ac0a48ad89e351c`. I verified the runtime paths rather than applying the suggestion verbatim: an immediate integer previously raised while being awaited, but an async provider could resolve to a truthy non-string and pass the old emptiness check. The adapter now distinguishes strings from awaitables and validates the resolved value as a non-empty string. The parameterized real-client regression covers both an empty string and an async provider resolving to `1`. Focused tests pass with OpenAI Python 2.18.0, the repository SDK, and 3.8.0 (16 each); the full unit suite passes (2439 passed, 69 skipped), and all-file pre-commit including strict mypy passes.

**Comment by IceCodeNew** [comment]
@coderabbitai full review\n\nPlease review exact head `26fd1522029c92c0476990172ac0a48ad89e351c` after the token-provider runtime validation fix.
